lizard_auth_server/views.py

```diff
@@ -304,3 +324,190 @@ def profile(self):
+        self.sso_key = sso_key
+
+        if not (hasattr(request, 'user') and request.user.is_authenticated()):
+            # User has not been authenticated yet: return login form.
+            return super(JWTView, self).get(request, *args, **kwargs)
+
+        # User already is authenticated: return token,
+        # but not before verifying his portal access!
+
+        portal = Portal.objects.get(sso_key=sso_key)
+
+        if not UserProfile.has_profile(request.user, portal):
+            reason = _('You do not have access to this portal.')
+            return HttpResponseBadRequest(reason, content_type='text/plain')
+
+        self.token = self.get_token(request.user, portal)
+        return self._response(None)
+
+    def post(self, request, *args, **kwargs):
+        """Process login form."""
+        self.redirect_to = request.POST.get('next', '')
+        self.sso_key = request.POST.get('portal', '')
+        return super(JWTView, self).post(request, *args, **kwargs)
+
+    def get_context_data(self, **kwargs):
+        """Return a dictionary with template context."""
+        # The `next` hidden input element in the login template is not managed
+        # by the form class. For that reason, it's added to the context here.
+        context = super(JWTView, self).get_context_data(**kwargs)
+        context['next'] = self.redirect_to
+        return context
+
+    def get_initial(self):
+        """Return a dictionary for prepopulating the form."""
+        # Initialize the hidden `portal` field with the public key.
+        initial = super(JWTView, self).get_initial()
+        initial['portal'] = self.sso_key
+        return initial
+
+    def get_success_url(self):
+        """Return the `next` URL with token."""
+        # URL to parts.
+        scheme, netloc, path, params, query, fragment = parse.urlparse(
+            self.redirect_to)
+        # Add (extra) query string parameter.
+        data = parse.parse_qs(query)
+        data['access_token'] = self.token
+        query = parse.urlencode(data, True)
+        # Parts to URL.
+        success_url = parse.urlunparse((
+            scheme, netloc, path, params, query, fragment))
+        return str(success_url)
+
+    def form_valid(self, form):
+        """Redirect the user to the `next` URL or return a JSON response.
+
+        At this point, the user has successfully been authenticated and
+        it has been checked that he has access to the portal. Log the
+        user in, so he can easily refresh tokens, without having to
+        fill out the form again.
+
+        """
+        user = form.get_user()
+        login(self.request, user)
+        portal = form.cleaned_data['portal']
+        self.token = self.get_token(user, portal)
+        return self._response(form)
+
+    def _response(self, form):
+        if self.redirect_to:
+            # Token as query parameter of redirect URL.
+            return super(JWTView, self).form_valid(form)
```

Comment:
Spaghetti. `.form_valid()` roept `_response()` aan en dat roept `.form_valid()` weer aan. Weliswaar van de superclass, maar dit is o.a. de reden dat ik denk dat er wat teveel in één view gemixt wordt qua gedrag (html vs. json response).